interpreter/text/parser.mly

```diff
@@ -198,8 +206,14 @@ text_list :
 /* Types */
 
 value_type_list :
-  | /* empty */ { [] }
-  | VALUE_TYPE value_type_list { $1 :: $2 }
+  | /* empty */ { fun c -> [] }
+  | VALUE_TYPE value_type_list { fun c -> $1 :: ($2 c) }
+/* TODO:
+  | LPAR REF NAT RPAR
+    { fun c -> ObjType ($3 c type_).it }
+*/
+  | LPAR REF VAR RPAR value_type_list
+    { fun c -> ObjType (type_ c ($3 @@ at ())) :: ($5 c) }
```

Comment:
Hm, the AST should reflect the input format, so in the syntax, a reference type should just contain a `var`. That would probably also solve your forward reference problem and avoid all the extra thunkification you introduced. And it might be necessary to support recursive types anyway. However, it might also mean that you need to pass in the type context in a couple of places.